ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2837,264 @@ public sealed class Bot : IAsyncDisposable {
+
+			if (badgePage == null) {
+				ArchiLogger.LogGenericWarning(Strings.WarningCouldNotCheckBadges);
+
+				return null;
+			}
+
+			byte maxPages = 1;
+			IElement? htmlNode = badgePage.SelectSingleNode("(//a[@class='pagelink'])[last()]");
+
+			if (htmlNode != null) {
+				string lastPage = htmlNode.TextContent;
+
+				if (string.IsNullOrEmpty(lastPage)) {
+					ArchiLogger.LogNullError(nameof(lastPage));
+
+					return null;
+				}
+
+				if (!byte.TryParse(lastPage, out maxPages) || (maxPages == 0)) {
+					ArchiLogger.LogNullError(nameof(maxPages));
+
+					return null;
+				}
+			}
+
+			HashSet<uint>? firstPageResult = GetPossiblyCompletedBadgeAppIDs(badgePage);
+
+			if (firstPageResult == null) {
+				return null;
+			}
+
+			switch (ASF.GlobalConfig?.OptimizationMode) {
+				case GlobalConfig.EOptimizationMode.MinMemoryUsage:
+					for (byte page = 2; page <= maxPages; page++) {
+						HashSet<uint>? pageIDs = await GetPossiblyCompletedBadgeAppIDs(page).ConfigureAwait(false);
+
+						if (pageIDs == null) {
+							return null;
+						}
+
+						firstPageResult.UnionWith(pageIDs);
+					}
+
+					return firstPageResult;
+				default:
+					HashSet<Task<HashSet<uint>?>> tasks = new HashSet<Task<HashSet<uint>?>>(maxPages - 1);
+
+					for (byte page = 2; page <= maxPages; page++) {
+						// We need a copy of variable being passed when in for loops, as loop will proceed before our task is launched
+						byte currentPage = page;
+						tasks.Add(GetPossiblyCompletedBadgeAppIDs(currentPage));
+					}
+
+					IList<HashSet<uint>?> results = await Utilities.InParallel(tasks).ConfigureAwait(false);
+
+					foreach (HashSet<uint>? result in results) {
+						if (result == null) {
+							return null;
+						}
+
+						firstPageResult.UnionWith(result);
+					}
+
+					return firstPageResult;
+			}
+		}
+
+		private async Task<HashSet<uint>?> GetPossiblyCompletedBadgeAppIDs(byte page) {
+			if (page == 0) {
+				throw new ArgumentNullException(nameof(page));
+			}
+
+			using IDocument? badgePage = await ArchiWebHandler.GetBadgePage(page).ConfigureAwait(false);
+
+			if (badgePage == null) {
+				ArchiLogger.LogGenericWarning(Strings.WarningCouldNotCheckBadges);
+
+				return null;
+			}
+
+			return GetPossiblyCompletedBadgeAppIDs(badgePage);
+		}
+
+		private HashSet<uint>? GetPossiblyCompletedBadgeAppIDs(IDocument badgePage) {
+			if (badgePage == null) {
+				throw new ArgumentNullException(nameof(badgePage));
+			}
+
+			List<IElement> linkElements = badgePage.SelectNodes("//a[@class='badge_craft_button']");
+
+			// We need to also select all badges that we have max level, as those will not display with a craft button
+			// Level 5 is maximum level for card badges according to https://steamcommunity.com/tradingcards/faq
+			linkElements.AddRange(badgePage.SelectNodes("//div[contains(@class, 'badge_row') and .//div[@class='badge_info_description']/div[contains(text(), 'Level 5')]]/a[@class='badge_row_overlay']"));
```

Comment:
I did not check for equality, as the text also contains tons of additional newlines, carriage-returns and tabs, as well as the info about how much XP was awarded for it. I didn't want to depend on chars that are used for formatting this mess (or whatever reason Valve has behind adding that).

Reply:
I don't mean the second part of the xpath, I've checked the page source - probably it's the only way to do it properly. I am talking about `//div[contains(@class, 'badge_row')` part, which means you get every div node from the document and check if `@class` contains `badge_row`. You can select `//div[@class='badges_sheet']` and then leave the rest XPath as it is, it will optimize node selecting by not iterating with `contains` through the entire document.

Reply:
Usually I'm after performance but in this case I agree with @Abrynos - if something is more bullet-proof against possible changes, we should go for more bullet-proof method as long as it has acceptable performance hit.

This has acceptable performance hit imho, of course it doesn't mean that it's the best solution and nothing can be done to further improve it, but unless there is solution with similar resistance against changes, we probably shouldn't change it.

Reply:
And when talking about possible improvements I'm talking for example about what @Vital7 suggested in his second message above - this indeed helps and doesn't hinder the resistance, we should do it if it makes sense.

Reply:
Oh yes. I'm with you on this one @Vital7